torch/csrc/dynamo/eval_frame.c

```diff
@@ -0,0 +1,578 @@
+  PyObject* old_callback = eval_frame_callback_get();
+
+  // owned by caller
+  Py_INCREF(old_callback);
+
+  if (old_callback != Py_None && new_callback == Py_None) {
+    decrement_working_threads(tstate);
+  } else if (old_callback == Py_None && new_callback != Py_None) {
+    increment_working_threads(tstate);
+  }
+
+  Py_INCREF(new_callback);
+  Py_DECREF(old_callback);
+
+  // Set thread local callback. This will drive behavior of our shim, if/when it
+  // is installed.
+  eval_frame_callback_set(new_callback);
+
+  return old_callback;
+}
+
+static PyObject* set_eval_frame_py(PyObject* dummy, PyObject* args) {
+  PyObject* callback = NULL;
+  if (!PyArg_ParseTuple(args, "O:callback", &callback)) {
+    DEBUG_TRACE0("arg error");
+    return NULL;
+  }
+  if (callback != Py_None && callback != Py_False &&
+      !PyCallable_Check(callback)) {
+    DEBUG_TRACE0("arg error");
+    PyErr_SetString(PyExc_TypeError, "expected a callable");
+    return NULL;
+  }
+  DEBUG_TRACE(
+      "python enabled=%d and is run_only=%d",
+      callback != Py_None,
+      callback == Py_False);
+  return set_eval_frame(callback, PyThreadState_GET());
+}
+
+static PyObject* reset_code(PyObject* dummy, PyObject* args) {
+  PyObject* code = NULL;
+  if (!PyArg_ParseTuple(args, "O:code", &code)) {
+    DEBUG_TRACE0("arg error");
+    return NULL;
+  }
+  if (!PyCode_Check(code)) {
+    DEBUG_TRACE0("arg error");
+    PyErr_SetString(PyExc_TypeError, "expected a code object");
+    return NULL;
+  }
+
+  destroy_cache_entry(get_extra((PyCodeObject*)code));
+  set_extra((PyCodeObject*)code, NULL);
+  Py_RETURN_NONE;
+}
+
+static PyObject* unsupported(PyObject* dummy, PyObject* args) {
+  // a dummy C function used in testing
+  PyObject* obj1 = NULL;
+  PyObject* obj2 = NULL;
+  if (!PyArg_ParseTuple(args, "OO", &obj1, &obj2)) {
+    return NULL;
+  }
+  Py_INCREF(obj2);
+  return obj2;
+}
+
+static PyObject* skip_code(PyObject* dummy, PyObject* args) {
+  PyObject* obj = NULL;
+  if (!PyArg_ParseTuple(args, "O", &obj)) {
+    return NULL;
+  }
+  if (!PyCode_Check(obj)) {
+    PyErr_SetString(PyExc_TypeError, "expected a code object");
+    return NULL;
+  }
+  set_extra((PyCodeObject*)obj, SKIP_CODE);
+  Py_RETURN_NONE;
+}
+
+static PyObject* set_guard_fail_hook(PyObject* dummy, PyObject* args) {
+  PyObject* obj = NULL;
+  if (!PyArg_ParseTuple(args, "O", &obj)) {
+    return NULL;
+  }
+  Py_XDECREF(guard_fail_hook);
+  if (obj == Py_None) {
+    guard_fail_hook = NULL;
+  } else {
+    guard_fail_hook = obj;
+    Py_INCREF(guard_fail_hook);
+  }
+  Py_RETURN_NONE;
+}
+
+static PyObject* set_guard_error_hook(PyObject* dummy, PyObject* args) {
+  PyObject* obj = NULL;
+  if (!PyArg_ParseTuple(args, "O", &obj)) {
+    return NULL;
+  }
+  Py_XDECREF(guard_error_hook);
+  if (obj == Py_None) {
+    guard_error_hook = NULL;
+  } else {
+    guard_error_hook = obj;
+    Py_INCREF(guard_error_hook);
+  }
+  Py_RETURN_NONE;
+}
+
+static PyMethodDef _methods[] = {
```

Comment:
All of these should have their content wrapped into HANDLE_TH_ERRORS / END_HANDLE_TH_ERRORS{_RET}
to ensure proper propagation of error and warnings.

Reply:
This only exists in C++

Reply:
We could move these bindings out of the c file though?